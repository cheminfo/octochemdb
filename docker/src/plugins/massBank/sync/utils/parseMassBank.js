import { Spectrum } from 'mass-tools';
import { xNormed, xyObjectToXY } from 'ml-spectra-processing';
import { parseMSP } from 'msp-parser';
import OCL from 'openchemlib';

import debugLibrary from '../../../../utils/Debug.js';
import { getNoStereosFromCache } from '../../../../utils/getNoStereosFromCache.js';

const debug = debugLibrary('parseMassBank');
export async function* parseMassBank(blob, connection) {
  try {
    const parsedData = parseMSP(blob);
    for (let data of parsedData) {
      let result = {};
      if (data.meta) {
        result._id = data.meta['DB#'];

        let oclMolecule = OCL.Molecule.fromSmiles(data.meta.SMILES);
        let ocl = await getNoStereosFromCache(oclMolecule, connection);
        result.data = {
          ocl,
        };
        let dataPeaks = { x: data.variables.x.data, y: data.variables.y.data };
        let spectrumToBeFilter = new Spectrum(dataPeaks);
        let minMaxX = spectrumToBeFilter.minMaxX();
        let slots = (minMaxX.max - minMaxX.min) / 0.1 - 1;
        xNormed(spectrumToBeFilter.data.y, {
          algorithm: 'max',
          output: spectrumToBeFilter.data.y,
        });
        let bestPeaks = spectrumToBeFilter.getBestPeaks({
          numberSlots: slots,
          numberCloseSlots: slots,
          limit: 100,
          threshold: 0.01,
        });
        let bestPeaksXY = xyObjectToXY(bestPeaks);
        result.data.spectrum = bestPeaksXY;
        result.data.spectrum.numberOfPeaks = bestPeaks.length;
        result.data.spectrum.instrument = data.meta.Instrument;
        result.data.spectrum.instrumentType = data.meta.Instrument_type;
        result.data.spectrum.ionMode = data.meta.Ion_mode;
        result.data.spectrum.adduct = data.meta.Spectrum_type;
        result.data.spectrum.collisionEnergy = data.meta.Collision_energy;
        result.data.spectrum.msLevel = data.meta.Spectrum_type.replace(
          /MS/,
          '',
        );
      }
      yield result;
    }
  } catch (e) {
    debug(e.message, {
      collection: 'massbank',
      connection,
      stack: e.stack,
    });
  }
}
