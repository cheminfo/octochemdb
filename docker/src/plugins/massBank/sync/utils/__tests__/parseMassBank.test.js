import { readFileSync } from 'fs';
import { join } from 'path';

import dotenv from 'dotenv';
import { test, expect } from 'vitest';

import { parseMassBank } from '../parseMassBank';

dotenv.config();
test(
  'parseMassBank',
  async () => {
    const blob = readFileSync(join(__dirname, 'data/massBank.msp'));
    const connection = 'test';
    const results = [];
    for await (const result of parseMassBank(blob, connection)) {
      results.push(result);
    }
    expect(results).toHaveLength(2);
    expect(results).toMatchSnapshot();
  },
  { timeout: 30000 },
);
