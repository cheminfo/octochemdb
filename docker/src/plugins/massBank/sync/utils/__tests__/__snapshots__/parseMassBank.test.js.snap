// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseMassBank 1`] = `
[
  {
    "_id": "MSBNK-AAFC-AC000854",
    "data": {
      "ocl": {
        "coordinates": "!Bb@K~@Ha}b@JH@ha}?\`BH?[\\\\Bb@K~_{]}bGw~_{\\\\BbOs~_xa}bGw~_p",
        "idCode": "fjm\`P@I@PIvQQIQZIKQPiTYdkTmUTp@QCd@@",
        "index": [
          255,
          255,
          255,
          255,
          171,
          247,
          254,
          254,
          74,
          123,
          215,
          188,
          216,
          52,
          218,
          170,
          148,
          138,
          136,
          180,
          32,
          9,
          98,
          95,
          36,
          66,
          108,
          177,
          141,
          2,
          180,
          198,
          32,
          37,
          17,
          120,
          166,
          0,
          163,
          26,
          128,
          16,
          53,
          8,
          128,
          22,
          5,
          57,
          0,
          0,
          16,
          144,
          41,
          20,
          0,
          0,
          0,
          0,
          96,
          8,
          64,
          0,
          8,
          0,
        ],
        "noStereoID": "fjm\`P@MHqpNRQYQYSIQQHtYdlsUUTC@H@A~X@",
        "noStereoTautomerID": "fjm\`P@A@pivRJJYSIIQRiLydmUUUUUUT@FjTXq\`fJXYcfAXUcVKX]cv@xcaNLxKbnOtafGFKGpq|L",
      },
      "spectrum": {
        "adduct": "MS2",
        "collisionEnergy": "30(NCE)",
        "data": {
          "x": [
            91.0542,
            125.0233,
            154.0499,
            155.0577,
            185.0961,
            200.107,
            229.0859,
            246.1125,
          ],
          "y": [
            0.24524524524524524,
            1,
            0.08008008008008008,
            0.35535535535535534,
            0.34934934934934936,
            0.04504504504504504,
            0.14214214214214213,
            0.7347347347347347,
          ],
        },
        "instrument": "Q-Exactive Orbitrap Thermo Scientific",
        "instrumentType": "LC-ESI-ITFT",
        "ionMode": "POSITIVE",
        "msLevel": "2",
        "numberOfPeaks": 8,
      },
    },
  },
  {
    "_id": "MSBNK-AAFC-AC000292",
    "data": {
      "ocl": {
        "coordinates": "!B?g~w_Xa}mpK~_xc|?\`Bw?[_}bOw~@Ha}bOrH?Hc|?\`BH?_x@bOrw@ox@bGt",
        "idCode": "fncAP@@HeFnQQJJJJHkIKJF\\\\UkTEQDMULTBJCh",
        "index": [
          247,
          127,
          255,
          255,
          189,
          254,
          185,
          239,
          110,
          153,
          244,
          27,
          9,
          13,
          154,
          83,
          17,
          206,
          147,
          213,
          98,
          178,
          48,
          8,
          133,
          192,
          4,
          0,
          20,
          7,
          48,
          130,
          0,
          72,
          88,
          8,
          26,
          64,
          96,
          68,
          0,
          4,
          82,
          97,
          128,
          32,
          32,
          64,
          32,
          2,
          16,
          176,
          0,
          18,
          2,
          74,
          16,
          0,
          64,
          4,
          4,
          128,
          192,
          17,
        ],
        "noStereoID": "fncAP@DXUEnRQSJJJIGIKIR\\\\VlpTERUURP@Gy\`@",
        "noStereoTautomerID": "fncAP@@XTdjTs\\\\|snk]oU[UUUUUUUT@FhTXqafNXeaVMXMavOXCbNDxs\`nFxQFX\\\\XL_CGpqZfOacxX",
      },
      "spectrum": {
        "adduct": "MS2",
        "collisionEnergy": "50(NCE)",
        "data": {
          "x": [
            55.0185,
            67.0542,
            69.0699,
            71.0491,
            79.0542,
            81.0699,
            83.0491,
            85.0648,
            91.0542,
            93.0699,
            95.0855,
            97.0648,
            105.0699,
            107.0855,
            115.0542,
            117.0699,
            119.0855,
            123.0441,
            125.0961,
            129.0699,
            131.0855,
            133.0648,
            141.0699,
            145.0648,
            147.0441,
            149.0597,
            157.0648,
            159.0804,
            160.0519,
            161.0597,
            163.039,
            165.0182,
            169.0648,
            173.0597,
            175.0754,
            177.0546,
            185.0597,
            187.0754,
            189.0546,
            197.0597,
            199.0754,
            201.091,
            203.0703,
            205.0859,
            211.0754,
            213.0546,
            214.0624,
            215.0703,
            217.0859,
            223.0754,
            225.091,
            226.0624,
            227.0703,
            228.0781,
            229.0859,
            231.0652,
            241.0859,
            243.1016,
            250.0988,
            255.138,
            265.1223,
            283.1329,
          ],
          "y": [
            0.11411411411411411,
            0.07607607607607608,
            0.8388388388388388,
            0.03203203203203203,
            0.062062062062062065,
            0.055055055055055056,
            0.04004004004004004,
            0.12212212212212212,
            0.0940940940940941,
            0.05105105105105105,
            0.056056056056056056,
            0.23723723723723725,
            0.07907907907907907,
            0.04704704704704705,
            0.05305305305305305,
            0.08108108108108109,
            0.05905905905905906,
            0.1031031031031031,
            0.04504504504504504,
            0.1111111111111111,
            0.13313313313313313,
            0.04004004004004004,
            0.2072072072072072,
            0.03903903903903904,
            0.07007007007007007,
            0.06806806806806806,
            0.45345345345345345,
            0.07807807807807808,
            0.06806806806806806,
            0.31431431431431434,
            0.06006006006006006,
            0.13013013013013014,
            0.11911911911911911,
            0.0980980980980981,
            0.3563563563563564,
            0.12812812812812813,
            1,
            0.8958958958958959,
            0.48148148148148145,
            0.06306306306306306,
            0.1061061061061061,
            0.07707707707707707,
            0.8028028028028028,
            0.12212212212212212,
            0.04204204204204204,
            0.15915915915915915,
            0.037037037037037035,
            0.11211211211211211,
            0.04104104104104104,
            0.05405405405405406,
            0.07307307307307308,
            0.03303303303303303,
            0.12112112112112113,
            0.036036036036036036,
            0.09109109109109109,
            0.055055055055055056,
            0.0890890890890891,
            0.04104104104104104,
            0.04604604604604605,
            0.03903903903903904,
            0.08608608608608609,
            0.044044044044044044,
          ],
        },
        "instrument": "Q-Exactive Orbitrap Thermo Scientific",
        "instrumentType": "LC-ESI-ITFT",
        "ionMode": "POSITIVE",
        "msLevel": "2",
        "numberOfPeaks": 62,
      },
    },
  },
]
`;
